rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,223 @@
+we cannot generate comprehensive [schema.org][] metadata for
+attachments: [see this comment][].
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the page body.  This inhibits creative use of our
+content.
+
+
+## Proposal
+
+In our content schemas we already have an [`asset_link` type][], which
+is used for attachments in specialist documents and manual sections.
+It's also used in travel advice pages for the page image and a single
+attachment.  This type solves all of our problems: [Publishing API can
+render such attachments with govspeak][], and content items can expose
+this data to enable programmatic use.
+
+However, the metadata Govspeak expects and the `asset_link` type are
+*informally* the same.  And furthermore, Whitehall has much richer
+attachment metadata than either `asset_link` or Govspeak allow.
+
+Given that `asset_link` is currently used for both images and
+attachments, adding all of the missing metadata to `asset_link` feels
+a mistake: images and attachments have different metadata, so we
+shouldn't conflate the two.
+
+I propose we do this:
+
+1. Split `asset_link` into two new types: `attachment_asset` and
+   `image_asset`, updating the schemas which use `asset_link`
+   accordingly.
+
+   **Migration concerns:** This can be done without breaking
+   compatibility.
+
+1. Add new optional fields for extra metadata:
+
+   - Attachments:
+     - `accessible`
+     - `alternative_format_contact_email`
+     - `command_paper_number`
+     - `file_size`
+     - `hoc_paper_number`
+     - `isbn`
+     - `locale`
+     - `number_of_pages`
+     - `parliamentary_session`
+     - `preview_url`
+     - `unique_reference`
+     - `unnumbered_command_paper`
+     - `unnumbered_hoc_paper`
+   - Images:
+     - `alt_text`
+     - `caption`
+     - `credit`
+     - `id`
+
+   I've left out the following fields because they can be deduced from
+   others: `csv?`, `external?`, `external_url`, `file_extension`, `html?`,
+   `opendocument?`, `pdf?`
+
+   And these fields because they don't seem to be used:
+   `print_meta_data_contact_address`, `web_isbn`.
+
+   And these fields because the Publishing Workflow team are planning
+   to remove them from Whitehall: `order_url`, `price_in_pence`.
+
+   And these fields because they don't need to be exposed outside the
+   publishing app: `attachable_id`, `attachable_type`,
+   `attachment_data_id`, `deleted`, `ordering`, `slug`.
+
+   **Migration concerns:** This can be done without breaking
+   compatibility.
+
+1. Replace `content_id` in attachments with `id`.
+
+    **Migration concerns:** this would need to be done in four steps:
+    add `id`; update apps; republish docs; remove `content_id`.
+
+1. Remove non-useful metadata fields:
+
+   - Attachments: `created_at` and `updated_at`
+   - Images: `content_type`
+
+   **Migration concerns:** this would need to be done in two steps:
```

Comment:
And specialist-publisher displays that metadata in a table: https://github.com/alphagov/specialist-publisher/blob/51cbb843af9650e3b4ef3648d232d57872418989/app/views/documents/show.html.erb#L76-L77

I guess it depends on how useful publishers find that information.  We could remove it, but if they use it (eg, to quickly check if someone's updated an attachment which needed changing) then we'd probably want to keep it, even though it may be confusing to end-users.

Reply:
Oh dear, that's a bit of a blocker - and reasonable since Specialist Publisher is trying to (ab)use Publishing API as a not-so-private database. Should we consider a(nother) separate schema for Specialist Publisher attachments, so we don't interfere with its domain model?

Reply:
I think this could actually be a good candidate for a "basic attachment" type, giving us this hierarchy of types:

- `base_attachment_asset` - has just the fields used by specialist-publisher
- `attachment_asset` - imports `base_attachment_asset` and adds the other fields in Kevin's post
- `extended_attachment_asset` - imports `attachment_asset` and adds the other Whitehall fields in Kevin's post

This would make the `created_at` and `updated_at` fields present but optional in all other attachment types.  I'd be wary of doing this:

```
      base_attachment
         |       |
attachment        specialist_attachment
```

(a fork in the type tree)

Or this:

```
attachment    specialist_attachment
```

(two totally disconnected types)

As I really don't want us to end up in the situation where our attachment types end up diverging and we get logic in govspeak like "if it's a regular attachment do *this*, otherwise if it's a specialist attachment do *that*."  I feel like, if we do have multiple attachment types, there's a clear hierarchy and each extends the one below it.

If we went for this I'd suggest removing `attachment_asset_list` and having the schemas for doctypes just have an appropriate list field.

What do you think? (cc @kevindew)

Reply:
@barrucadu I agree we should have a simple hierarchy. I do think we should consider something custom for Specialist Publisher, because it is custom and prone to diverge, being coupled to the internal needs of the app. I'd call it a smell.

The nice thing about `extended_attachment_asset` is that it implicitly disallows further subclassing. Reading the name, I'm interpreting it as "a dump of extra stuff that we don't care about much". I wonder if we should consider something more specific, that's equally "final". The idea that comes to mind is a `government_attachment`, since (I think) that uniquely groups all the HOC/ref/etc. fields we're discussing - not sure if @kevindew meant to duplicate these fields between the types he was suggesting?

Reply:
Oh great catch on the updated_at / created_at for Specialist Publisher, that is a pain.

That does seem to suggest multiple attachment types unless we're just prepared to keep those fields everywhere else.

So yes if we proceed with an approach towards other types what we'd likely do is a have a base_attachment set up in our schemas that acts like an abstract class. I'd guess that'd have:

accessible, alternative_format_contact_email, content_type, filename, file_size, locale, number_of_pages, preview_url, title (required), and url (required)

A `specialist_publisher_attachment_asset` would inherit from `base_attachment_asset` that and add the following fields:

content_id (required), created_at, updated_at

A `attachment_asset` would inherit from `base_attachment_asset` and have the following fields:

id (required)

A `government_attachemnt_asset` (or `publication_attachment_asset`) would inherit from `attachment_asset` and have the following fields:

attachment_type (required), command_paper_number, hoc_paper_number, isbn, parliamentary_session, unique_reference, unnumbered_command_paper, unnumbered_hoc_paper

A couple of things to note are:

- There isn't going to be a signifier as to which of these types you'd be dealing with based on the data that would depend on an applications knowledge that a document type follows a particular schema
- As attachment_type would only be present on `government_attachment_asset` we'd have to assume that attachments without it are file attachments, which seems a pretty reasonable assumption. This would probably only be a factor for govspeak that doesn't know about document types.

Would we be able to switch any apps that use the existing asset_link away from that so they can use the `attachment_asset` or would we find other apps end up needing to use the specialist publisher variant?

This also includes my suggestions on the ever thorny problem of names :-) I think for specialist publisher we might want to just verbosely name it to avoid any confusion - as a specialist attachment sounds like it's a special type rather than for a specialist publisher. I prefer the idea of naming the publication ones government than extended as that feels more direct, though I have concerns that it may cause confusion as all attachments are government ones. I slightly prefer using the name publication myself as these are government publications, even when they're on the consultation document type. Though I can understand resistance based on document type confusion so I'd be willing to go with government if that's what's preferred.

Reply:
Regarding other apps, manuals-publisher [sets `updated_at` and `created_at`](https://github.com/alphagov/manuals-publisher/blob/master/app/adapters/publishing_adapter.rb#L247-L256).  But uses mongo as its database, so those fields could be dropped without impacting the experience for publishers.

```ruby
        attachments: section.attachments.map do |attachment|
          {
            content_id: SecureRandom.uuid,
            title: attachment.title,
            url: attachment.file_url,
            updated_at: attachment.updated_at,
            created_at: attachment.created_at,
            content_type: attachment.content_type,
          }
        end,
```

And travel-advice-publisher [doesn't set them at all](https://github.com/alphagov/travel-advice-publisher/blob/master/app/presenters/asset_presenter.rb#L10-L17).

```ruby
  def present
    if asset
      {
        "url" => asset["file_url"],
        "content_type" => asset["content_type"],
      }
    end
  end
```

So I think we could change those to use `attachment_asset`, and have both of them use `SecureRandom.uuid` for the `id`.

Reply:
Since it seems like I'm the only one who wants specialist publisher attachments to be the base type, I'll go for the separate type approach.  I'll write up that change now.